File: 03 - MatchScheduler/src/clients/TeamGeneratorClient.ts
import axios, { AxiosInstance } from 'axios';

/**
 * Team data structure from TeamGenerator API
 */
export interface TeamData {
  readonly id: string;
  readonly name: string;
}

/**
 * Player data structure from TeamGenerator API
 */
export interface PlayerData {
  readonly id: string;
  readonly name: string;
  readonly position: string;
  readonly skillLevel: number;
}

/**
 * Match data structure from TeamGenerator API
 */
export interface MatchData {
  readonly id: string;
  readonly homeTeamId: string;
  readonly awayTeamId: string;
  readonly scheduledDate: string;
  readonly venue: string;
  readonly status: string;
  readonly createdAt: string;
  readonly notes?: string;
}

/**
 * Client for communicating with the TeamGenerator service
 */
export class TeamGeneratorClient {
  private readonly httpClient: AxiosInstance;

  constructor(baseUrl: string = process.env.TEAM_GENERATOR_URL || 'http://localhost:5252') {
    this.httpClient = axios.create({
      baseURL: baseUrl,
      timeout: 5000,
      headers: {
        'Content-Type': 'application/json'
      }
    });
  }

  /**
   * Get all teams from TeamGenerator
   */
  async getAllTeams(): Promise<TeamData[]> {
    try {
      const response = await this.httpClient.get<TeamData[]>('/api/TeamsData');
      return response.data;
    } catch (error) {
      console.error('Failed to fetch teams from TeamGenerator:', error);
      throw new Error('Unable to fetch teams from TeamGenerator service');
    }
  }

  /**
   * Get a specific team by ID
   */
  async getTeamById(id: string): Promise<TeamData | null> {
    try {
      const response = await this.httpClient.get<TeamData>(`/api/TeamsData/${id}`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        return null;
      }
      console.error(`Failed to fetch team ${id} from TeamGenerator:`, error);
      throw new Error(`Unable to fetch team ${id} from TeamGenerator service`);
    }
  }

  /**
   * Get all players from TeamGenerator
   */
  async getAllPlayers(): Promise<PlayerData[]> {
    try {
      const response = await this.httpClient.get<PlayerData[]>('/api/Players');
      return response.data;
    } catch (error) {
      console.error('Failed to fetch players from TeamGenerator:', error);
      throw new Error('Unable to fetch players from TeamGenerator service');
    }
  }

  /**
   * Get a specific player by ID
   */
  async getPlayerById(id: string): Promise<PlayerData | null> {
    try {
      const response = await this.httpClient.get<PlayerData>(`/api/Players/${id}`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        return null;
      }
      console.error(`Failed to fetch player ${id} from TeamGenerator:`, error);
      throw new Error(`Unable to fetch player ${id} from TeamGenerator service`);
    }
  }

  /**
   * Get all matches from TeamGenerator
   */
  async getAllMatches(): Promise<MatchData[]> {
    try {
      const response = await this.httpClient.get<MatchData[]>('/api/Matches');
      return response.data;
    } catch (error) {
      console.error('Failed to fetch matches from TeamGenerator:', error);
      throw new Error('Unable to fetch matches from TeamGenerator service');
    }
  }

  /**
   * Get a specific match by ID
   */
  async getMatchById(id: string): Promise<MatchData | null> {
    try {
      const response = await this.httpClient.get<MatchData>(`/api/Matches/${id}`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        return null;
      }
      console.error(`Failed to fetch match ${id} from TeamGenerator:`, error);
      throw new Error(`Unable to fetch match ${id} from TeamGenerator service`);
    }
  }

  /**
   * Get matches for a specific team
   */
  async getMatchesForTeam(teamId: string): Promise<MatchData[]> {
    try {
      const response = await this.httpClient.get<MatchData[]>(`/api/Matches/team/${teamId}`);
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        return [];
      }
      console.error(`Failed to fetch matches for team ${teamId} from TeamGenerator:`, error);
      throw new Error(`Unable to fetch matches for team ${teamId} from TeamGenerator service`);
    }
  }
}

// Singleton instance
export const teamGeneratorClient = new TeamGeneratorClient();

File: 02 - BettingService/src/clients/team_generator_client.py
import requests
from typing import List, Optional, Dict, Any
import os
from dataclasses import dataclass
from datetime import datetime

@dataclass
class TeamData:
    id: str
    name: str

@dataclass
class PlayerData:
    id: str
    name: str
    position: str
    skill_level: int

@dataclass
class MatchData:
    id: str
    home_team_id: str
    away_team_id: str
    scheduled_date: datetime
    venue: str
    status: str
    created_at: datetime
    notes: Optional[str] = None

class TeamGeneratorClient:
    """Client for communicating with the TeamGenerator service"""
    
    def __init__(self, base_url: str = None):
        self.base_url = base_url or os.getenv('TEAM_GENERATOR_URL', 'http://localhost:5252')
        self.session = requests.Session()
        self.session.headers.update({'Content-Type': 'application/json'})
        self.timeout = 5
    
    def get_all_teams(self) -> List[TeamData]:
        """Get all teams from TeamGenerator"""
        try:
            response = self.session.get(f"{self.base_url}/api/TeamsData", timeout=self.timeout)
            response.raise_for_status()
            data = response.json()
            return [TeamData(id=team['id'], name=team['name']) for team in data]
        except requests.RequestException as e:
            print(f"Failed to fetch teams from TeamGenerator: {e}")
            raise Exception("Unable to fetch teams from TeamGenerator service")
    
    def get_team_by_id(self, team_id: str) -> Optional[TeamData]:
        """Get a specific team by ID"""
        try:
            response = self.session.get(f"{self.base_url}/api/TeamsData/{team_id}", timeout=self.timeout)
            if response.status_code == 404:
                return None
            response.raise_for_status()
            data = response.json()
            return TeamData(id=data['id'], name=data['name'])
        except requests.RequestException as e:
            print(f"Failed to fetch team {team_id} from TeamGenerator: {e}")
            raise Exception(f"Unable to fetch team {team_id} from TeamGenerator service")
    
    def get_all_matches(self) -> List[MatchData]:
        """Get all matches from TeamGenerator"""
        try:
            response = self.session.get(f"{self.base_url}/api/Matches", timeout=self.timeout)
            response.raise_for_status()
            data = response.json()
            matches = []
            for match in data:
                matches.append(MatchData(
                    id=match['id'],
                    home_team_id=match['homeTeamId'],
                    away_team_id=match['awayTeamId'],
                    scheduled_date=datetime.fromisoformat(match['scheduledDate'].replace('Z', '+00:00')),
                    venue=match['venue'],
                    status=match['status'],
                    created_at=datetime.fromisoformat(match['createdAt'].replace('Z', '+00:00')),
                    notes=match.get('notes')
                ))
            return matches
        except requests.RequestException as e:
            print(f"Failed to fetch matches from TeamGenerator: {e}")
            raise Exception("Unable to fetch matches from TeamGenerator service")
    
    def get_match_by_id(self, match_id: str) -> Optional[MatchData]:
        """Get a specific match by ID"""
        try:
            response = self.session.get(f"{self.base_url}/api/Matches/{match_id}", timeout=self.timeout)
            if response.status_code == 404:
                return None
            response.raise_for_status()
            data = response.json()
            return MatchData(
                id=data['id'],
                home_team_id=data['homeTeamId'],
                away_team_id=data['awayTeamId'],
                scheduled_date=datetime.fromisoformat(data['scheduledDate'].replace('Z', '+00:00')),
                venue=data['venue'],
                status=data['status'],
                created_at=datetime.fromisoformat(data['createdAt'].replace('Z', '+00:00')),
                notes=data.get('notes')
            )
        except requests.RequestException as e:
            print(f"Failed to fetch match {match_id} from TeamGenerator: {e}")
            raise Exception(f"Unable to fetch match {match_id} from TeamGenerator service")

# Singleton instance
team_generator_client = TeamGeneratorClient()